File: dlox/lib/src/resolver.dart
import 'expression.dart';
import 'interpreter.dart';
import 'lox.dart';
import 'nil.dart';
import 'statement.dart';
import 'token.dart';
import 'visitor.dart';

enum _FunctionType { NONE, FUNCTION, METHOD, INITIALIZER }

enum _ClassType { NONE, CLASS, SUBCLASS }

class Resolver implements ExprVisitor<void>, StmtVisitor<void> {
  final Interpreter _interpreter;
  final _scopes = <Map<String, bool>>[];

  var _currentFunction = _FunctionType.NONE;
  var _currentClass = _ClassType.NONE;

  Resolver(this._interpreter);

  @override
  void visitAssignExpr(AssignExpr expr) {
    _resolveExpr(expr.value);
    _resolveLocal(expr, expr.name);
  }

  @override
  void visitBinaryExpr(BinaryExpr expr) {
    _resolveExpr(expr.left);
    _resolveExpr(expr.right);
  }

  @override
  void visitBlockStmt(BlockStmt stmt) {
    _beginScope();
    resolve(stmt.statements);
    _endScope();
  }

  @override
  void visitCallExpr(CallExpr expr) {
    _resolveExpr(expr.callee);
    expr.arguments.forEach(_resolveExpr);
  }

  @override
  void visitClassStmt(ClassStmt stmt) {
    final enclosingClass = _currentClass;
    _currentClass = _ClassType.CLASS;

    _define(stmt.name);

    final superclass = stmt.superclass;

    if (superclass != null && stmt.name.lexeme == superclass.name.lexeme) {
      Lox.error(superclass.name, "A class can't inherit from itself.");
    }

    if (superclass != null) {
      _currentClass = _ClassType.SUBCLASS;
      _resolveExpr(superclass);
    }

    if (superclass != null) {
      _beginScope();
      _scopes.last['super'] = true;
    }

    _beginScope();

    _scopes.last['this'] = true;

    stmt.methods.forEach((m) {
      final declaration = m.name.lexeme == 'init'
          ? _FunctionType.INITIALIZER
          : _FunctionType.METHOD;

      _resolveFunction(m, declaration);
    });

    _endScope();

    if (superclass != null) _endScope();

    _currentClass = enclosingClass;
  }

  @override
  void visitExpressionStmt(ExpressionStmt stmt) =>
      _resolveExpr(stmt.expression);

  @override
  void visitFunctionStmt(FunctionStmt stmt) {
    _declare(stmt.name);
    _define(stmt.name);
    _resolveFunction(stmt, _FunctionType.FUNCTION);
  }

  @override
  void visitGetExpr(GetExpr expr) => _resolveExpr(expr.object);

  @override
  void visitGroupingExpr(GroupingExpr expr) => _resolveExpr(expr.expression);

  @override
  void visitIfStmt(IfStmt stmt) {
    _resolveExpr(stmt.condition);
    _resolveStmt(stmt.thenBranch);
    if (stmt.elseBranch != null) {
      _resolveStmt(stmt.elseBranch!);
    }
  }

  @override
  void visitLiteralExpr(LiteralExpr expr) {}

  @override
  void visitLogicalExpr(LogicalExpr expr) {
    _resolveExpr(expr.left);
    _resolveExpr(expr.right);
  }

  @override
  void visitNil(Nil expr) {}

  @override
  void visitPrintStmt(PrintStmt stmt) => _resolveExpr(stmt.expression);

  @override
  void visitReturnStmt(ReturnStmt stmt) {
    if (_currentFunction == _FunctionType.NONE) {
      Lox.error(stmt.keyword, "Can't return from top-level code.");
    }
    if (stmt.value is! Nil) {
      if (_currentFunction == _FunctionType.INITIALIZER) {
        Lox.error(stmt.keyword, "Can't return a value from an initializer.");
      }
      _resolveExpr(stmt.value);
    }
  }

  @override
  void visitSetExpr(SetExpr expr) {
    _resolveExpr(expr.value);
    _resolveExpr(expr.object);
  }

  @override
  void visitSuperExpr(SuperExpr expr) {
    if (_currentClass == _ClassType.NONE) {
      Lox.error(expr.keyword, "Can't use 'super' outside of a class.");
    } else if (_currentClass != _ClassType.SUBCLASS) {
      Lox.error(
          expr.keyword, "Can't use 'super' in a class with no superclass.");
    }
    _resolveLocal(expr, expr.keyword);
  }

  @override
  void visitThisExpr(ThisExpr expr) {
    if (_currentClass == _ClassType.NONE) {
      Lox.error(expr.keyword, "Can't use 'this' outside of a class.");
      return;
    }

    _resolveLocal(expr, expr.keyword);
  }

  @override
  void visitUnaryExpr(UnaryExpr expr) => _resolveExpr(expr.right);

  @override
  void visitVarStmt(VarStmt stmt) {
    _declare(stmt.name);
    if (stmt.initializer != null) {
      _resolveExpr(stmt.initializer!);
    }
    _define(stmt.name);
  }

  @override
  void visitVariableExpr(VariableExpr expr) {
    if (_scopes.isNotEmpty && _scopes.last[expr.name.lexeme] == false) {
      Lox.error(expr.name, "Can't read local variable in its own initializer.");
    }
    _resolveLocal(expr, expr.name);
  }

  @override
  void visitWhileStmt(WhileStmt stmt) {
    _resolveExpr(stmt.condition);
    _resolveStmt(stmt.body);
  }

  void resolve(List<Stmt> statements) => statements.forEach(_resolveStmt);

  void _resolveStmt(Stmt stmt) => stmt.accept(this);

  void _resolveExpr(Expr expr) => expr.accept(this);

  void _resolveLocal(Expr expr, Token name) {
    for (var i = _scopes.length - 1; i >= 0; i--) {
      if (_scopes[i].containsKey(name.lexeme)) {
        _interpreter.resolve(expr, _scopes.length - 1 - i);
        return;
      }
    }
  }

  void _resolveFunction(FunctionStmt func, _FunctionType type) {
    final enclosingFunction = _currentFunction;
    _currentFunction = type;

    _beginScope();
    func.params.forEach((param) {
      _declare(param);
      _define(param);
    });
    resolve(func.body);
    _endScope();

    _currentFunction = enclosingFunction;
  }

  void _beginScope() => _scopes.add(<String, bool>{});

  void _endScope() => _scopes.removeLast();

  void _declare(Token name) {
    if (_scopes.isNotEmpty) {
      final scope = _scopes.last;
      if (scope.containsKey(name.lexeme)) {
        Lox.error(name, 'Already variable with this name in this scope.');
      }
      scope[name.lexeme] = false;
    }
  }

  void _define(Token name) {
    if (_scopes.isNotEmpty) {
      _scopes.last[name.lexeme] = true;
    }
  }
}

File: rlox/src/resolver.rs
use crate::ast::{BinOp, Expr, Keyword, Lit, Stmt, UnOp};
use crate::visitor::{ExprVisitor, StmtVisitor};
use crate::{Interpreter, LoxError};
use std::collections::HashMap;

pub struct Resolver<'a> {
    interpreter: &'a mut Interpreter,
    scopes: Vec<HashMap<String, bool>>,
}

impl<'a> Resolver<'a> {
    #[allow(unused)]
    pub fn new(interpreter: &mut Interpreter) -> Resolver {
        Resolver {
            interpreter,
            scopes: Vec::new(),
        }
    }

    fn begin_scope(&mut self) {
        self.scopes.push(HashMap::new())
    }

    pub fn resolve(&mut self, stmts: &[Stmt]) -> Result<(), LoxError> {
        for stmt in stmts {
            self.resolve_stmt(stmt)?
        }

        Ok(())
    }

    fn resolve_stmt(&mut self, stmt: &Stmt) -> Result<(), LoxError> {
        self.visit_stmt(stmt)
    }

    fn resolve_expr(&mut self, expr: &Expr) -> Result<(), LoxError> {
        self.visit_expr(expr)
    }

    fn resolve_function(&mut self, params: &[String], body: &[Stmt]) -> Result<(), LoxError> {
        self.begin_scope();

        for param in params {
            self.declare(param);
            self.define(param);
        }

        self.resolve(body)?;
        self.end_scope();

        Ok(())
    }

    fn end_scope(&mut self) {
        self.scopes.pop();
    }

    fn resolve_local(&mut self, name: &str) -> Result<(), LoxError> {
        for (i, scope) in self.scopes.iter().enumerate() {
            if scope.contains_key(name) {
                self.interpreter.resolve(name, self.scopes.len() - 1 - i);
                break;
            }
        }

        Ok(())
    }

    fn declare(&mut self, name: &str) {
        if let Some(scope) = self.scopes.last_mut() {
            scope.insert(name.into(), false);
        }
    }

    fn define(&mut self, name: &str) {
        if let Some(scope) = self.scopes.last_mut() {
            scope.insert(name.into(), true);
        }
    }
}

impl StmtVisitor for Resolver<'_> {
    type Output = Result<(), LoxError>;

    fn visit_block_stmt(&mut self, stmts: &[Stmt]) -> Self::Output {
        self.begin_scope();
        self.resolve(stmts)?;
        self.end_scope();

        Ok(())
    }

    fn visit_expression_stmt(&mut self, expr: &Expr) -> Self::Output {
        self.resolve_expr(expr)
    }

    fn visit_function_stmt(
        &mut self,
        name: &str,
        params: &[String],
        body: &[Stmt],
        _line: u64,
    ) -> Self::Output {
        self.declare(name);
        self.define(name);
        self.resolve_function(params, body)
    }

    fn visit_if_stmt(
        &mut self,
        condition: &Expr,
        then: &Stmt,
        r#else: Option<&Stmt>,
    ) -> Self::Output {
        self.resolve_expr(condition)?;
        self.resolve_stmt(then)?;

        if let Some(stmt) = r#else {
            self.resolve_stmt(stmt)?
        }
        Ok(())
    }

    fn visit_print_stmt(&mut self, expr: &Expr) -> Self::Output {
        self.resolve_expr(expr)
    }

    fn visit_return_stmt(&mut self, expr: Option<&Expr>, _: u64) -> Self::Output {
        if let Some(exp) = expr {
            self.resolve_expr(exp)?
        }

        Ok(())
    }

    fn visit_var_stmt(&mut self, name: &str, initializer: Option<&Expr>) -> Self::Output {
        self.declare(name);
        if let Some(expr) = initializer {
            self.resolve_expr(expr)?;
        }
        self.define(name);

        Ok(())
    }

    fn visit_while_stmt(&mut self, condition: &Expr, body: &Stmt) -> Self::Output {
        self.resolve_expr(condition)?;
        self.resolve_stmt(body)
    }
}

impl ExprVisitor for Resolver<'_> {
    type Output = Result<(), LoxError>;

    fn visit_assign_expr(&mut self, name: &str, expr: &Expr, _: u64) -> Self::Output {
        self.resolve_expr(expr)?;
        self.resolve_local(name)
    }

    fn visit_binary_expr(&mut self, lhs: &Expr, _: BinOp, rhs: &Expr, _: u64) -> Self::Output {
        self.resolve_expr(lhs)?;
        self.resolve_expr(rhs)
    }

    fn visit_call_expr(&mut self, callee: &Expr, args: &[Expr], _: u64) -> Self::Output {
        self.resolve_expr(callee)?;
        for arg in args {
            self.resolve_expr(arg)?;
        }
        Ok(())
    }

    fn visit_literal_expr(&mut self, _: &Lit) -> Self::Output {
        Ok(())
    }

    fn visit_logical_expr(&mut self, lhs: &Expr, _: Keyword, rhs: &Expr, _: u64) -> Self::Output {
        self.resolve_expr(lhs)?;
        self.resolve_expr(rhs)
    }

    fn visit_unary_expr(&mut self, _: UnOp, rhs: &Expr, _: u64) -> Self::Output {
        self.resolve_expr(rhs)
    }

    fn visit_variable_expr(&mut self, name: &str, _: u64) -> Self::Output {
        if let Some(false) = self.scopes.last().and_then(|scope| scope.get(name)) {
            let msg = "Can't read variable in its own initializer.";
            return Err(LoxError::Compile(msg.into()));
        }

        self.resolve_local(name)
    }
}
